### PR TITLE
REST layer refactoring/cleanup

### PR DESCRIPTION
**Pull request by lmsurpre**
1. added originalRequestUri to FHIRRequestContext
2. set it from FHIRRestServletFilter
3. use it from FHIRResource
4. remove unused `queryString` parm from SearchUtil.parseQueryParameters

Signed-off-by: Lee Surprenant <lmsurpre@us.ibm.com>

### CODE REVIEW

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java [hunk 1]
```diff
@@ -3560,10 +3561,9 @@ private String serializeOperationOutcome(OperationOutcome oo) {
         }
     }
 
-    private synchronized CapabilityStatement getCapabilityStatement() throws Exception {
+    private synchronized CapabilityStatement getCapabilityStatement() throws FHIROperationException {
```
Comment (prb112) [inline fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: After our earlier discussion, this is only called once? or does it synchronize across multiple requests, I could see this being slow unintentionally.

File: fhir-server/src/main/java/com/ibm/fhir/server/filter/rest/FHIRRestServletFilter.java [hunk 2]
```diff
@@ -39,54 +41,54 @@
 /**
  * This class is a servlet filter which is registered with the REST API's servlet. The main purpose of the class is to
  * log entry/exit information and elapsed time for each REST API request processed by the server.
- * 
- * @author padams
  */
-public class FHIRRestServletFilter implements Filter {
+public class FHIRRestServletFilter extends HttpFilter {
+    private static final long serialVersionUID = 1L;
```
Comment (prb112) [inline fhir-server/src/main/java/com/ibm/fhir/server/filter/rest/FHIRRestServletFilter.java]: are you actually serializing this class?  best to generate it not use 1l.

File: fhir-server/src/main/java/com/ibm/fhir/server/filter/rest/FHIRRestServletFilter.java [hunk 3]
```diff
@@ -100,56 +102,33 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
...
-                }
-            }
-            FHIRRequestContext.get().setReturnPreference(returnPref);
+            HTTPReturnPreference returnPref = computeReturnPref(request, handlingPref);
+            context.setReturnPreference(returnPref);
 
             // Pass the request through to the next filter in the chain.
             chain.doFilter(request, response);
-        } catch (FHIRException e) {
+        } catch (Exception e) {
```
Comment (prb112) [inline fhir-server/src/main/java/com/ibm/fhir/server/filter/rest/FHIRRestServletFilter.java]: Totally agree with this change!

File: fhir-server/src/main/java/com/ibm/fhir/server/filter/rest/FHIRRestServletFilter.java [hunk 4]
```diff
@@ -39,54 +41,54 @@
...
  */
-public class FHIRRestServletFilter implements Filter {
+public class FHIRRestServletFilter extends HttpFilter {
+    private static final long serialVersionUID = 1L;
+
     private static final Logger log = Logger.getLogger(FHIRRestServletFilter.class.getName());
 
     private static String tenantIdHeaderName = null;
     private static String datastoreIdHeaderName = null;
+    private static String originalRequestUriHeaderName = null;
```
Comment (prb112) [inline fhir-server/src/main/java/com/ibm/fhir/server/filter/rest/FHIRRestServletFilter.java]: I thought Filters were stateless? ... so the reason I mention this... 
if the configuration changes, a server restart needs to be called to re-activate the init filter

Comment (prb112) [reply fhir-server/src/main/java/com/ibm/fhir/server/filter/rest/FHIRRestServletFilter.java]: maybe document that behavior?

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java (hunk 1 above)
Comment (lmsurpre) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: I don't think this comment is really related to this change.  It was (and still is) called on each `metadata` request and thats what we have #734 for

File: fhir-server/src/main/java/com/ibm/fhir/server/filter/rest/FHIRRestServletFilter.java (hunk 2 above)
Comment (lmsurpre) [reply fhir-server/src/main/java/com/ibm/fhir/server/filter/rest/FHIRRestServletFilter.java]: I am not, it was just to make eclipse not complain about missing it (apparently HttpFilter is serializable whereas Filter was not)

File: fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java (hunk 1 above)
Comment (prb112) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: yeah ... I'm just wondering if we're injecting a pause with the synchronized when two threads are trying to generate it

Comment (prb112) [reply fhir-server/src/main/java/com/ibm/fhir/server/resources/FHIRResource.java]: No need to resolve here

File: fhir-server/src/main/java/com/ibm/fhir/server/filter/rest/FHIRRestServletFilter.java (hunk 4 above)
Comment (lmsurpre) [reply fhir-server/src/main/java/com/ibm/fhir/server/filter/rest/FHIRRestServletFilter.java]: added to section 5.1.3 of the userguide

**Comment by lmsurpre** [comment]
discussed with paul and we think 1L is fine for now...no immediate plans to serialize it

**Pull request merged**